flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java

```diff
@@ -0,0 +1,310 @@
+	public <K> KeyedStateBackend<K> getKeyedStateBackend() {
+		return (KeyedStateBackend<K>) keyedStateBackend;
+	}
+
+	public OperatorStateBackend getOperatorStateBackend() {
+		return operatorStateBackend;
+	}
+
+	public <N, S extends State, T> S getOrCreateKeyedState(
+			TypeSerializer<N> namespaceSerializer,
+			StateDescriptor<S, T> stateDescriptor) throws Exception {
+
+		if (keyedStateStore != null) {
+			return keyedStateBackend.getOrCreateKeyedState(namespaceSerializer, stateDescriptor);
+		}
+		else {
+			throw new IllegalStateException("Cannot create partitioned state. " +
+					"The keyed state backend has not been set." +
+					"This indicates that the operator is not partitioned/keyed.");
+		}
+	}
+
+	/**
+	 * Creates a partitioned state handle, using the state backend configured for this task.
+	 *
+	 * @throws IllegalStateException Thrown, if the key/value state was already initialized.
+	 * @throws Exception Thrown, if the state backend cannot create the key/value state.
+	 */
+	protected <S extends State, N> S getPartitionedState(
+			N namespace,
+			TypeSerializer<N> namespaceSerializer,
+			StateDescriptor<S, ?> stateDescriptor) throws Exception {
+
+		/*
+	    TODO: NOTE: This method does a lot of work caching / retrieving states just to update the namespace.
+	    This method should be removed for the sake of namespaces being lazily fetched from the keyed
+	    state backend, or being set on the state directly.
+	    */
+
+		if (keyedStateStore != null) {
+			return keyedStateBackend.getPartitionedState(namespace, namespaceSerializer, stateDescriptor);
+		} else {
+			throw new RuntimeException("Cannot create partitioned state. The keyed state " +
+				"backend has not been set. This indicates that the operator is not " +
+				"partitioned/keyed.");
+		}
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public void setCurrentKey(Object key) {
+		if (keyedStateBackend != null) {
+			try {
+				// need to work around type restrictions
+				@SuppressWarnings("unchecked,rawtypes")
+				AbstractKeyedStateBackend rawBackend = (AbstractKeyedStateBackend) keyedStateBackend;
+
+				rawBackend.setCurrentKey(key);
+			} catch (Exception e) {
+				throw new RuntimeException("Exception occurred while setting the current key context.", e);
+			}
+		}
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public Object getCurrentKey() {
+		if (keyedStateBackend != null) {
+			return keyedStateBackend.getCurrentKey();
+		} else {
+			throw new UnsupportedOperationException("Key can only be retrieved on KeyedStream.");
+		}
+	}
+
+	public Optional<KeyedStateStore> getKeyedStateStore() {
+		return Optional.ofNullable(keyedStateStore);
+	}
+
+	/**
+	 * Custom state handling hooks to be invoked by {@link StreamOperatorStateHandler}.
+	 */
+	public interface CheckpointedStreamOperator {
+		void initializeState(StateInitializationContext context) throws Exception;
+
+		void snapshotState(StateSnapshotContext context) throws Exception;
+	}
```

Comment:
This interface is implemented by `@PublicEvolving AbstractStreamOperator` (which is de-facto API).
Should it be a top-level `@PublicEvolving` class then? With Javadoc moved from `AbstractStreamOperator` into it.

Reply:
My intention was to make this interface `@Internal`, and making (keeping as they are) those methods `@PublicEvolving` only on the `AbstractStreamOperator` level. Maybe it should be turned into `@PublicEvolving` interface at some point, but I didn't want to commit us to it right now.